src/Adapter/MSTest.CoreAdapter/Execution/TestClassInfo.cs

```diff
@@ -298,19 +340,26 @@ public void RunClassInitialize(TestContext testContext)
         /// <returns>
         /// Any exception that can be thrown as part of a class cleanup as warning messages.
         /// </returns>
-        [SuppressMessageAttribute("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "Requirement is to handle all kinds of user exceptions and message appropriately.")]
+        [SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "Requirement is to handle all kinds of user exceptions and message appropriately.")]
         public string RunClassCleanup()
         {
-            if (this.ClassCleanupMethod == null)
+            if (this.ClassCleanupMethod is null && !this.BaseClassInitAndCleanupMethods.Any(p => p.Item2 != null))
             {
                 return null;
             }
 
-            if (this.IsClassInitializeExecuted || this.ClassInitializeMethod == null)
+            if (this.IsClassInitializeExecuted || this.ClassInitializeMethod is null || this.ExecutedBaseClassInitializeMethods.Any())
             {
+                var classCleanupMethod = this.classCleanupMethod;
                 try
                 {
-                    this.ClassCleanupMethod.InvokeAsSynchronousTask(null);
+                    classCleanupMethod?.InvokeAsSynchronousTask(null);
+                    var baseClassCleanupQueue = new Queue<Tuple<MethodInfo, MethodInfo>>(this.BaseClassInitAndCleanupMethods);
+                    while (baseClassCleanupQueue.Count > 0)
+                    {
+                        classCleanupMethod = baseClassCleanupQueue.Dequeue().Item2;
+                        classCleanupMethod?.InvokeAsSynchronousTask(null);
```

Comment:
shouldn't we be only executing cleanups whose inits have run? This seems to be running all cleanups.

Reply:
Maybe we could just push Cleanups that need to be executed into a Stack instead of `ExecutedBaseClassInitializeMethods`

Reply:
Yup, I had a block of code that I removed right before pushing the latest changes, handling the logic for deciding those cleanup method that will be executed. But, now that you mentioned that last part, I think that'd be better. For that, renamed the property you mentioned, now it's called: `BaseClassCleanupMethodsStack `not sure if the name is 100% ok, and in the `RunClassInitialize `method, I'll be pushing the cleanups into that stack to then execute them in the `RunClassCleanup `method.

Reply:
Sounds good. It doesn't look like we need a Queue here. We could just pop the methods off the stack..